keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,305 @@
+    * Only CSINode containing `spec.drivers[xyz].allocatable` for a volume plugin (to simulate kubelet after deprecation period).
+    * Only Node containing `status.allocatable` for a volume plugin (to simulate old kubelet).
+  * Test results should be ideally the same as before the KEP.
+    * Any deviation needs to be approved by sig-scheduling.
+
+* Run e2e tests and kubelet version skew tests to check that scheduler picks the right values from CSINode or Node.
+
+* Add e2e test that runs pods with both in-tree volumes and CSI driver for the same storage backend and check that they share the same volume limits.
+
+### Graduation Criteria
+
+##### Alpha -> Beta Graduation
+
+N/A (`AttachVolumeLimit` feature is already beta).
+
+##### Beta -> GA Graduation
+
+It must graduate together with CSI migration.
+
+### Upgrade / Downgrade / Version Skew Strategy
+
+During upgrade, downgrade or version skew, kubelet may be older that scheduler. Kubelet will not fill `CSINode.spec` with volume limits and it will fill volume limits into `Node.status`. Scheduler must fall back to `Node.status` when `CSINode` is not available or its `spec` does not contain a volume plugin / CSI driver.
+
+#### Interaction with CSI migration
+
+In-tree volume plugins are being migrated to CSI in a [separate KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190129-csi-migration.md). This section covers possible situations:
+
+* CSI migration feature is `off` for a volume plugin and CSI driver is not installed on a node:
+  * In-tree volume plugin is used for in-tree PVs.
+  * Kubelet gets volume limits from the `Node` object and existing predicate do their work.
+
+* CSI migration feature is `off` for a volume plugin and CSI driver (for the same storage backend) is installed on a node:
+  * In-tree volume plugin is used for in-tree PVs.
+  * CSI driver is used for CSI PVs.
+  * Kubelet gets volume limits from the in-tree volume plugin and updates `Node` object with limit. For in-tree volumes existing predicate to their work.
+  * When CSI driver is registered, kubelet gets its volume limit through CSI and updates `CSINode`. At this point existing predicate for in-tree plugin becomes NOOP and both in-tree PVs and CSI PVs are counted by `MaxCSIVolumeCountPred` predicate.
+
+* CSI migration feature is `on` for a volume plugin and there is no CSI driver installed yet:
+  * In-tree volume plugin is "off".
+  * CSI driver is used both for in-tree and CSI PVs (if the driver is installed).
+  * Kubelet creates `CSINode` during node registration with empty `spec.drivers`.
+    * Scheduler does not put pods with CSI volumes for the CSI driver there, because the driver is missing in `CSINode.spec.drivers`.
+  * When CSI driver is registered, kubelet gets its volume limit through CSI and adds the driver to `CSINode.spec.drivers` .
```

Comment:
Clarify here that after the driver is installed, then the scheduler can start scheduling pods to that node

Reply:
fixed.